### ISSUE TITLE
Structure learning with start/initial DAG

### ISSUE THREAD

**Issue by MathijsPost** [pgmpy/pgmpy#1356]
### Subject of the issue
I am trying to estimate the structure of a DAG representing a dataset. Initially, this took quite long so I created quite an extensive model myself to minimize the search space (so I thought). I am running into runtime issues, but the syntax in my code does not match the documentation. 

### Your environment
* pgmpy version = 0.1.11
* Python version = 3.7.4
* Operating System = Windows 10

### Steps to reproduce
```
model = BayesianModel([('A', 'B'), \
                        ('B', 'C'), \
                        ('B', 'D'), \ 
                       .... some more edges...)])

hc = HillClimbSearch(data, scoring_method=BicScore(data))
best_model = hc.estimate(start = model)
print(best_model.edges())
```

This is the code I am using. 'data' is a pandas dataset that I decrease to only 100 (or even 10) rows to speed up the process. I am using a model I created by hand as the 'start_dag', however somehow my pgmpy does not recognize the command 'start_dag', instead I have to use 'start'. Additionally, the commands 'max_iter' and 'epsilon' do not seem to do anything. I set 'max_iter' equal to 10 and even 1 having no effect. 'fixed_edges' is also not recognized by python.

This does run without errors, but it takes very long. Yesterday it showed no results after 12 hours, but if I use estimate without an initial DAG with a 100-row dataset it takes 2 minutes. 

How can it be that pgmpy does not recognize commands from the documentation such as 'start_dag', or that 'max_iter' has no effect? Am I missing something?

**Comment by ankurankan** [comment]
@MathijsPost The online documentation is for v0.1.12, that's why some of the arguments are different/not working. A lot of things in Hill Climb were changed in the last release, and should also be faster now. Could you give the new version a try and let me know if you face the same issues?

Also, just a pointer that the `scoring_method` argument has also been moved to the `estimate` method now, so make sure to change that in your code as it fails silently and uses the default method otherwise.

**Comment by MathijsPost** [comment]
@ankurankan Thank you for your clear and swift reply. I will try to do so! I am having some difficulties, have you ever run into a similar issue using Anaconda prompt? 

InvalidArchiveError('Error with archive C:\\Users\\TUDelft SID\\Anaconda3\\pkgs\\pgmpy-0.1.12-py37_0.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nCould not unlink')

**Comment by ankurankan** [comment]
@MathijsPost Maybe you have already tried this but running `conda clean -a` before installing might fix it. Another thing you can try is to completely remove the previous version and do a fresh install instead of trying to update it. If this doesn't fix it, please let me know as I have updated the conda package this morning, so there might be some problem with the actual uploaded package.

**Comment by MathijsPost** [comment]
@ankurankan I just tried that, but no luck. The download/extraction bar stays at 0%

Downloading and Extracting Packages
pgmpy-0.1.12         | 9.6 MB    |                                                                              |   0%

InvalidArchiveError('Error with archive C:\\Users\\TUDelft SID\\Anaconda3\\pkgs\\pgmpy-0.1.12-py37_0.tar.bz2.  You probably need to delete and re-download or re-create this file.  Message from libarchive was:\n\nCould not unlink')

**Comment by ankurankan** [comment]
@MathijsPost Not sure why this is happening. But you can also try installing through pip using: `pip install pgmpy`

**Comment by MathijsPost** [comment]
@ankurankan Thanks! While anaconda doesn't show it, it seems updated as start_dag works! Much appreciated!

**Comment by Cyx123hn** [comment]
how to use it in multi-label learning @ankurankan

**Comment by ankurankan** [comment]
@Cyx123hn Sorry, I am not sure what exactly do you mean by multi-label learning. If you mean variables with multiple classes, everything should work as it is. Also, if it's unrelated to this issue, can you please create a new issue for your question.

**Comment by Cyx123hn** [comment]
Yes, that's what I mean. If I use this for multi label learning, how to deal with it's feature? Is the input of structural learning just classes？




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "pgmpy/pgmpy"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年11月30日(星期一) 晚上8:49
收件人:&nbsp;"pgmpy/pgmpy"<pgmpy@noreply.github.com&gt;;
抄送:&nbsp;"npcjaiot"<1060074099@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [pgmpy/pgmpy] Structure learning with start/initial DAG (#1356)





 
@Cyx123hn Sorry, I am not sure what exactly do you mean by multi-label learning. If you mean variables with multiple classes, everything should work as it is. Also, if it's unrelated to this issue, can you please create a new issue for your question.

**Comment by ankurankan** [comment]
@Cyx123hn The structure learning algorithms take dataset as input and based on the dataset it will automatically infer the number of labels.

**Comment by MathijsPost** [comment]
@ankurankan I have a follow-up question, everything seems to work, however by using either a `start_dag` or `fixed_edges` iterations do not seem to happen. I am using roughly 25 variables as input, and as initial DAG/edges I add about 30. After one hour no iterations are finished. First I assumed that adding so many edges could speed up the process, however this does not seem to be the case. Is my input for the estimate with `HillClimbSearch` too complex or detailed?

**Comment by MathijsPost** [comment]
I used `bicscore` at first, `k2score` does seem to work!

**Comment by MathijsPost** [comment]
`bdeuscore` seems to work too! Any reason why `bicscore` does not start the iterations?

**Comment by ankurankan** [comment]
@MathijsPost I can't think of a reason for this. Would it be possible for you to share some minimal reproducible code, so that I can try it on my end?

**Comment by MathijsPost** [comment]
Hi @ankurankan , apologies for the late reply. Today I changed some values and value-types. At first the dataset was a combination of categorical values and integers, but today I changed everything to integers which works much faster. However, this does raise the question, does pgmpy interpret integers the same way as categorical values? That is, is 2 always larger than 1, or does pgmpy interpret these as categories anyway?

**Comment by ankurankan** [comment]
@MathijsPost Interesting, I expected the runtime to be the same for both integers or strings. HillClimb considers everything as unordered categorical. So, if you provide it with integral values, it will consider each value to be a category, and the value itself wouldn't matter (i.e. changing the value for the category wouldn't make any difference).

**Comment by MathijsPost** [comment]
@ankurankan What could have caused the issue, was that I used pandas qcut() to discretize the data in bins. However, such a bin/category is still a float, which could have led to the issue. If anything comes up I will keep you posted.

**Comment by ankurankan** [comment]
@MathijsPost Since you mentioned that you discretized your variables, were you using continuous data before? Because in that case, the algorithm assumes each value to be a category which would result in a huge number of categories and slow down the algorithm.

**Comment by MathijsPost** [comment]
@ankurankan I discretized the variables before I opened this issue. As far as I know I never used continuous data. I just tried the structure learning with a dataset of 500 rows and a maximum of 10 iterations which worked flawless. However, more iterations and at some point more rows in the dataset leads to my RAM memory being used to the maximum. Spyder says it uses 99% of my memory. Possibly the unique values for some variables are too large. Is there a guideline for this? Some variables have 4 states and a few up to 10. However, there are 2 variables with 100+ states, could that be the issue?
